### PR TITLE
Items endpoints

### PR DESCRIPTION
**Pull request by audrius-savickas**

### CODE REVIEW

File: Domain/Entities/Review.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+public class Review : BaseEntity
+{
+    public Item Item { get; set; }
+    public int ItemId { get; set; }
+    public Employee Employee { get; set; }
+    public int EmployeeId { get; set; }
+    public int Rating { get; set; }
+    public String? Description { get; set; }
+    public String? Photo { get; set; }
```
Comment (aauga) [inline Domain/Entities/Review.cs]: panaikink nullables

File: WebApi/Controllers/ItemController.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /// </summary>
+    [HttpPut("{id}")]
+    public async Task<ActionResult<ItemDto>> Update(int id, [FromBody] ItemBodyDto itemDto)
+    {
+        return await Mediator.Send(new UpdateItemCommand(id, itemDto));
+    }
+
+    /// <summary>
+    /// Delete item.
+    /// </summary>
```
Comment (aauga) [inline WebApi/Controllers/ItemController.cs]: atnaujinau sita vieta del warningu, naudok [Summary(string)] atributa

File: WebApi/Properties/launchSettings.json [hunk 3]
```diff
@@ -11,7 +11,6 @@
   "profiles": {
     "http": {
       "commandName": "Project",
-      "dotnetRunMessages": true,
```
Comment (aauga) [inline WebApi/Properties/launchSettings.json]: ka tai keicia? :D

File: Application/Items/Commands/CreateItemCommand.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+public record CreateItemCommand : IAuthorizedRequest<ItemDto>
+{
+    public Employee Employee;
+    public string? Title { get; init; }
+    public string? Category { get; init; }
+    public decimal Price { get; init; }
+    public string? Description { get; init; }
+    public string? Brand { get; init; }
+    public string? Photo { get; init; }
```
Comment (aauga) [inline Application/Items/Commands/CreateItemCommand.cs]: panaikink nullables

File: Application/Items/Commands/CreateItemCommand.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        _dbContext.Items.Add(entity);
+
+        try
+        {
+            await _dbContext.SaveChangesAsync();
+        }
+        catch (DbUpdateException)
+        {
+            throw new ForbiddenAccessException();
+        }
```
Comment (aauga) [inline Application/Items/Commands/CreateItemCommand.cs]: sito nereikia

File: Application/Items/Commands/CreateItemCommand.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        var entity = new Item
+        {
+            Title = request.Title,
+            Category = request.Category,
+            Price = request.Price,
+            Description = request.Description,
+            Brand = request.Brand,
+            Photo = request.Photo,
+            TenantId = (int)request.Employee.TenantId,
```
Comment (aauga) [inline Application/Items/Commands/CreateItemCommand.cs]: nereikia (int)

File: Application/Items/Commands/CreateItemCommand.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    public string? Category { get; init; }
+    public decimal Price { get; init; }
+    public string? Description { get; init; }
+    public string? Brand { get; init; }
+    public string? Photo { get; init; }
+
+    public async Task<bool> Authorize(Employee employee, IUserService userService, IApplicationDbContext dbContext)
+    {
+        this.Employee = employee;
+        return await userService.CanCreateItemAsync(employee);
```
Comment (aauga) [inline Application/Items/Commands/CreateItemCommand.cs]: CanManageTenant()

File: Application/Items/Commands/DeleteItemCommand.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+using MediatR;
+using Microsoft.EntityFrameworkCore;
+
+namespace Application.Items;
+
+public record DeleteItemCommand(int Id) : IAuthorizedRequest
+{
+    public async Task<bool> Authorize(Employee employee, IUserService userService, IApplicationDbContext dbContext)
+    {
+        return await userService.CanManageItemAsync(employee, Id);
```
Comment (aauga) [inline Application/Items/Commands/DeleteItemCommand.cs]: canManageTenant

File: Application/Items/Commands/DeleteItemCommand.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        _dbContext.Items.Remove(entity);
+
+        try
+        {
+            await _dbContext.SaveChangesAsync();
+        }
+        catch (DbUpdateException)
+        {
+            throw new ForbiddenAccessException();
+        }
```
Comment (aauga) [inline Application/Items/Commands/DeleteItemCommand.cs]: nereikia

File: Application/Items/Commands/UpdateItemCommand.cs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        entity.Photo = request.Item.Photo;
+
+        try
+        {
+            await _dbContext.SaveChangesAsync();
+        }
+        catch (DbUpdateException)
+        {
+            throw new ForbiddenAccessException();
+        }
```
Comment (aauga) [inline Application/Items/Commands/UpdateItemCommand.cs]: nereikia

File: Application/Items/Commands/UpdateItemCommand.cs [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+using MediatR;
+using Microsoft.EntityFrameworkCore;
+
+namespace Application.Items;
+
+public record UpdateItemCommand(int Id, ItemBodyDto Item) : IAuthorizedRequest<ItemDto>
+{
+    public async Task<bool> Authorize(Employee employee, IUserService userService, IApplicationDbContext dbContext)
+    {
+        return await userService.CanManageItemAsync(employee, Id);
```
Comment (aauga) [inline Application/Items/Commands/UpdateItemCommand.cs]: canmanagetenant

File: Application/Items/Queries/GetItemQuery.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+namespace Application.Items;
+
+public record GetItemQuery(int Id) : IAuthorizedRequest<ItemDto>
+{
+    public Employee Employee;
+
+    public async Task<bool> Authorize(Employee employee, IUserService userService, IApplicationDbContext dbContext)
+    {
+        this.Employee = employee;
+        return await userService.CanViewItemsAsync(employee, Id);
```
Comment (aauga) [inline Application/Items/Queries/GetItemQuery.cs]: canviewtenant

File: Infrastructure/Identity/UserService.cs [hunk 13]
```diff
@@ -56,4 +56,41 @@ private async Task<bool> TenantExistsAsync(int tenantId)
...
+        var authorizedTenant = item != null ? item.TenantId == employee.TenantId : true;
+        var authorizedPosition = employee.Position >= PositionType.Manager;
+
+        return await Task.FromResult(authorizedTenant && authorizedPosition);
+    }
+
+    private async Task<bool> ItemExistsAsync(int itemId)
+    {
+        return await _dbContext.Items.AnyAsync(x => x.Id == itemId);
+    }
```
Comment (aauga) [inline Infrastructure/Identity/UserService.cs]: uztenka canviewtenant canmanagetenant

File: WebApi/Properties/launchSettings.json (hunk 3 above)
Comment (audrius-savickas) [reply WebApi/Properties/launchSettings.json]: honestly idk, auto-generated gal del macOS, netycia ipushinau, will remove

File: Application/Items/Commands/CreateItemCommand.cs (hunk 7 above)
Comment (audrius-savickas) [reply Application/Items/Commands/CreateItemCommand.cs]: wait why? How does it relate to managing tenants?

File: Application/Items/Commands/DeleteItemCommand.cs (hunk 8 above)
Comment (audrius-savickas) [reply Application/Items/Commands/DeleteItemCommand.cs]: this is how the method would look like then if we want to have the same authorization logic:
```suggestion
        var item = await dbContext.Items.FindAsync(Id);
        return item != null ? await userService.CanManageTenantAsync(employee, item.TenantId) : true;
```
wouldn't it be nicer to have a separate `userService` method `CanManageItemAsync()` then?

File: Application/Items/Commands/CreateItemCommand.cs (hunk 7 above)
Comment (aauga) [reply Application/Items/Commands/CreateItemCommand.cs]: perhaps the naming is not the best, but CanViewTenant is for regular employees (f.e. cashiers), whereas CanManageTenant is for managers and admins. a manager should be able to edit, create, delete all items in a tenant (and CanManageTenant is fully enough for that), so I don't see where its helpful to have another function which does the same thing?

File: Application/Items/Commands/DeleteItemCommand.cs (hunk 8 above)
Comment (aauga) [reply Application/Items/Commands/DeleteItemCommand.cs]: why im against it: user service would be filled up with functions which would be used only in few places and few times, whereas canviewtenant and canmanagetenant can be used in every command. f.e. you'll use the canviewitem only in Items queries or commands

personally, I'd write the logic in the authorize method instead

File: Application/Items/Commands/CreateItemCommand.cs (hunk 7 above)
Comment (audrius-savickas) [reply Application/Items/Commands/CreateItemCommand.cs]: `CanManageTenant` expects `tenantId`, but when creating a new item, there's no `tenantId` to check, so it wouldn't work

Comment (audrius-savickas) [reply Application/Items/Commands/CreateItemCommand.cs]: So I guess I should still create `CanCreateTenant` method which only checks that employee has position manager or admin?

File: Application/Items/Commands/DeleteItemCommand.cs (hunk 8 above)
Comment (audrius-savickas) [reply Application/Items/Commands/DeleteItemCommand.cs]: ok then, I previously thought that logic in Authorize method would be bad, but if it's fine, i'll rewrite it there

File: Application/Items/Commands/CreateItemCommand.cs (hunk 7 above)
Comment (audrius-savickas) [reply Application/Items/Commands/CreateItemCommand.cs]: rewrote Create Item endpoint to take `tenantId` from path

**Review by krtrinkunas — APPROVED** [review]

**Review by aauga — APPROVED** [review]

**Pull request merged**